contracts/swaps/connectors/SwapsImplUniswapV3_ETH.sol

```diff
@@ -0,0 +1,369 @@
+
+    function dexSwap(
+        address sourceTokenAddress,
+        address destTokenAddress,
+        address receiverAddress,
+        address returnToSenderAddress,
+        uint256 minSourceTokenAmount,
+        uint256 maxSourceTokenAmount,
+        uint256 requiredDestTokenAmount,
+        bytes memory payload
+    )
+        public
+        returns (uint256 destTokenAmountReceived, uint256 sourceTokenAmountUsed)
+    {
+        require(sourceTokenAddress != destTokenAddress, "source == dest");
+        require(
+            supportedTokens[sourceTokenAddress] &&
+                supportedTokens[destTokenAddress],
+            "invalid tokens"
+        );
+
+        IERC20 sourceToken = IERC20(sourceTokenAddress);
+        address _thisAddress = address(this);
+        (sourceTokenAmountUsed, destTokenAmountReceived) = _swapWithUni(
+            sourceTokenAddress,
+            destTokenAddress,
+            receiverAddress,
+            minSourceTokenAmount,
+            maxSourceTokenAmount,
+            requiredDestTokenAmount,
+            payload
+        );
+
+        if (
+            returnToSenderAddress != _thisAddress &&
+            sourceTokenAmountUsed < maxSourceTokenAmount
+        ) {
+            // send unused source token back
+            sourceToken.safeTransfer(
+                returnToSenderAddress,
+                maxSourceTokenAmount - sourceTokenAmountUsed
+            );
+        }
+    }
+
+    function dexExpectedRate(
+        address sourceTokenAddress,
+        address destTokenAddress,
+        uint256 sourceTokenAmount
+    ) public view returns (uint256 expectedRate) {
+        revert("unsupported");
+    }
+
+    function dexAmountOut(bytes memory route, uint256 amountIn)
+        public
+        returns (uint256 amountOut, address midToken)
+    {
+        if (amountIn == 0) {
+            amountOut = 0;
+        } else if (amountIn != 0) {
+            amountOut = _getAmountOut(amountIn, route);
+        }
+    }
+
+    function dexAmountOutFormatted(bytes memory payload, uint256 amountIn)
+        public
+        returns (uint256 amountOut, address midToken)
+    {
+        IUniswapV3SwapRouter.ExactInputParams[] memory exactParams = abi.decode(
+            payload,
+            (IUniswapV3SwapRouter.ExactInputParams[])
+        );
+        uint256 totalAmounts = 0;
+        for (
+            uint256 uniqueInputParam = 0;
+            uniqueInputParam < exactParams.length;
+            uniqueInputParam++
+        ) {
+            totalAmounts = totalAmounts.add(
+                exactParams[uniqueInputParam].amountIn
+            );
+        }
+        if (totalAmounts < amountIn) {
+            exactParams[0].amountIn = exactParams[0].amountIn.add(
+                amountIn.sub(totalAmounts)
+            ); //adds displacement to first swap set
+        }
+        uint256 tempAmountOut;
+        for (uint256 i = 0; i < exactParams.length; i++) {
```

Comment:
this for can be extracted into a function. its used twice

Reply:
different call on the if statements.